File: backend/app/schemas/tender.py
"""Pydantic schemas for tender lookup and watchlist endpoints.

Design notes (03-CONTEXT.md, Decision 3):
- WatchlistAddRequest uses Field(min_length, max_length) + whitespace strip.
- NO regex on numberAnno — SPIKE-01 confirmed format "{digits}-{digits}" but
  variations are possible. Gate is: non-empty ≤ 100 chars after strip.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field, field_validator


class LotItem(BaseModel):
    model_config = {"from_attributes": True}

    lot_number: Optional[str] = None
    name_ru: Optional[str] = None
    name_kz: Optional[str] = None
    description_ru: Optional[str] = None
    amount: Optional[Any] = None


class TenderResponse(BaseModel):
    model_config = {"from_attributes": True}

    number_anno: str
    name_ru: Optional[str] = None
    name_kz: Optional[str] = None
    total_sum: Optional[Any] = None
    customer_name_ru: Optional[str] = None
    customer_name_kz: Optional[str] = None
    # refBuyStatusId: 220 = "Опубликовано (прием заявок)" / PublishedOrderTaking (SPIKE-01)
    status_id: Optional[int] = None
    status_name_ru: Optional[str] = None
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    publish_date: Optional[datetime] = None
    lots_data: Optional[list[Any]] = None
    cached_at: datetime


class WatchlistAddRequest(BaseModel):
    number_anno: str = Field(min_length=1, max_length=100)

    @field_validator("number_anno")
    @classmethod
    def strip_whitespace(cls, v: str) -> str:
        """Strip leading/trailing whitespace from the tender number."""
        stripped = v.strip()
        if not stripped:
            raise ValueError("Номер тендера не может быть пустым")
        return stripped


class WatchlistEntryResponse(BaseModel):
    model_config = {"from_attributes": True}

    tender: TenderResponse
    notification_on: bool
    added_at: datetime

File: frontend/src/types/tender.ts
/**
 * TypeScript types for the TenderIt tender-lookup feature.
 *
 * Key casing notes:
 * - Tender fields use snake_case (Pydantic default serialization from backend).
 * - lots_data items use camelCase (raw goszakup GraphQL response stored as JSONB).
 *   SPIKE-01 confirmed: lotNumber, nameRu, nameKz, descriptionRu, amount.
 */

/** A single lot from the goszakup TrdBuy response (camelCase — raw JSONB). */
export interface Lot {
  id?: number
  lotNumber?: string | null
  nameRu?: string | null
  nameKz?: string | null
  descriptionRu?: string | null
  /** count = количество единиц (quantity to supply). Confirmed 2026-07-23 via introspection. */
  count?: number | null
  /** amount = запланированная сумма (total budget). NOT quantity — do not use for price calc. */
  amount?: number | null
  refLotStatusId?: number | null
}

/** TenderResponse from GET /api/tenders/{number_anno} (snake_case). */
export interface Tender {
  number_anno: string
  name_ru: string | null
  name_kz: string | null
  total_sum: number | null
  customer_name_ru: string | null
  customer_name_kz: string | null
  /** refBuyStatusId: 220 = "Опубликовано (прием заявок)" (SPIKE-01) */
  status_id: number | null
  status_name_ru: string | null
  start_date: string | null
  end_date: string | null
  publish_date: string | null
  lots_data: Lot[] | null
  cached_at: string
}

/** WatchlistEntryResponse from POST /api/watchlist and GET /api/watchlist. */
export interface WatchlistEntry {
  tender: Tender
  notification_on: boolean
  added_at: string
}
